### PR TITLE
[API Compatibility] Implement paddle.addcmul API -part

### PR DESCRIPTION
**Pull request by Manfredss**
### PR Category
User Experience

### PR Types
New features

### Description

This PR implements the `addcmul` operator for PaddlePaddle, which performs element-wise multiplication of two tensors, multiplies the result by a scalar value, and adds it to an input tensor.

**Formula**: `output = input + value * tensor1 * tensor2`

This operator provides users with a convenient operation for combined multiply-add computations.

### Core Components

1. **C++ Kernels** (`paddle/phi/kernels/`)
   - Forward kernels: `addcmul_kernel.h` and implementations for CPU/GPU
   - Backward kernels: `addcmul_grad_kernel.h` and implementations for CPU/GPU
   - Implementation files in `impl/` directory with templated functions for different ranks (0-6D)

2. **Operator Configuration** (`paddle/phi/ops/yaml/`)
   - Added `addcmul` operator definition in `ops.yaml`
   - Added `addcmul_grad` backward operator in `backward.yaml`
   - Configured with proper infer_meta and kernel functions

3. **Shape Inference** (`paddle/phi/infermeta/`)
   - Implemented `AddcmulInferMeta` in `ternary.cc/h`
   - Handles broadcasting for three input tensors
   - Validates dimension compatibility

4. **PIR Support** (`paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/`)
   - Added `AddcmulOpInferSymbolicShape` for new IR system
   - Handles symbolic shape inference with broadcasting

5. **Python API** (`python/paddle/tensor/`)
   - Added `paddle.addcmul()` function in `math.py`
   - Registered `Tensor.addcmul()` method in `__init__.py`
   - Supports both dynamic and static graph modes

6. **Testing** (`test/legacy_test/`)
   - Comprehensive test suite with 52 test cases
   - Tests multiple data types: float16, float32, float64, bfloat16
   - Tests various tensor shapes and broadcasting scenarios
   - Tests gradient computation for all inputs
   - Tests zero-size tensors and error conditions
   - Tests both OpTest framework and high-level API

7. **Configuration** (`test/white_list/`)
   - Added `addcmul` to FP64 gradient threshold whitelist

## Features

-  **Multi-device support**: CPU and GPU (CUDA)
-  **Multiple data types**: float16, float32, float64, bfloat16
-  **Broadcasting**: Full NumPy-style broadcasting support
-  **Gradient support**: Automatic differentiation for all three inputs
-  **Tensor dimensions**: Supports 0D to 6D tensors
-  **API compatibility**: Similar interface to PyTorch's `torch.addcmul`
-  **Zero-size tensors**: Properly handles edge cases

## Testing Results

All 52 tests pass successfully:

```
(paddle) D:\Xue\ML\Paddle\PaddleDebug>python test/legacy_test/test_addcmul.py
WARNING: Logging before InitGoogleLogging() is written to STDERR
W0113 14:12:09.673414 20628 gpu_resources.cc:116] Please NOTE: device: 0, GPU Compute Capability: 12.0, Driver API Version: 13.1, Runtime API Version: 12.9
....I0113 14:12:09.695261 20628 pir_interpreter.cc:1529] New Executor is Running ...
I0113 14:12:09.695261 20628 pir_interpreter.cc:1552] pir interpreter is running by multi-thread mode ...
..I0113 14:12:09.702877 20628 program_interpreter.cc:255] New Executor is Running.
I0113 14:12:09.704878 20628 interpreter_util.cc:624] Standalone Executor is Used.
W0113 14:12:09.744876 20628 eager_utils.cc:3584] Paddle static graph(PIR) not support input out tensor for now!!!!!
C:\Users\***\anaconda3\envs\paddle\Lib\site-packages\paddle\pir\math_op_patch.py:241: UserWarning: Tensor do not have 'place' interface for pir graph mode, try not to use it. None will be returned.
  warnings.warn(
..............................................
----------------------------------------------------------------------
Ran 52 tests in 16.889s

OK
```

Test coverage includes:
- Basic functionality with various shapes (1D, 2D, 3D, large tensors)
- Different `value` parameters (positive, negative, default)
- Multiple data types (FP16, FP32, FP64, BF16)
- Broadcasting scenarios
- Gradient checks for all inputs
- Zero-size tensor edge cases
- Error handling for invalid inputs
- Both static and dynamic graph modes
- Tensor method (`tensor.addcmul()`)

### Dynamic Graph Mode
```python
import paddle

input = paddle.ones([2, 2])
tensor1 = paddle.ones([2, 2]) * 2
tensor2 = paddle.ones([2, 2]) * 3

# Using function API
out = paddle.addcmul(input, tensor1, tensor2, value=0.5)
# Result: [[4., 4.], [4., 4.]]

# Using tensor method
out = input.addcmul(tensor1, tensor2, value=0.5)
```

### Static Graph Mode
```python
import paddle

paddle.enable_static()
input = paddle.static.data('input', shape=[2, 2], dtype='float32')
tensor1 = paddle.static.data('tensor1', shape=[2, 2], dtype='float32')
tensor2 = paddle.static.data('tensor2', shape=[2, 2], dtype='float32')
out = paddle.addcmul(input, tensor1, tensor2, value=0.5)
```

### Broadcasting
```python
input = paddle.ones([3, 4])
tensor1 = paddle.randn([1, 4])
tensor2 = paddle.randn([3, 1])
out = paddle.addcmul(input, tensor1, tensor2, value=2.0)
```

## Backward Compatibility

This PR adds new functionality without modifying existing APIs or behaviors. It is fully backward compatible.

## Checklist

- [x] Implemented forward and backward kernels
- [x] Added operator YAML configurations
- [x] Implemented shape inference (InferMeta)
- [x] Added PIR symbolic shape inference
- [x] Created Python API wrapper
- [x] Registered tensor method
- [x] Added comprehensive test suite
- [x] All tests passing (52/52)
- [x] Added to gradient threshold whitelist
- [x] Code follows PaddlePaddle style guidelines
- [x] All comments in English
- [x] No linter errors

## Related Issues

[【启航计划】PaddlePaddle API兼容性增强 No.354](https://github.com/PaddlePaddle/Paddle/issues/76301)

## Additional Notes

- The operator uses Eigen for efficient computation with automatic vectorization
- Mixed precision computation is handled via `MPTypeTrait` for numerical stability
- Broadcasting follows NumPy semantics
- Gradient computation is mathematically verified and tested

## Files Changed

**New Files** (9):
```bash
paddle/phi/kernels/addcmul_kernel.h
paddle/phi/kernels/addcmul_grad_kernel.h
paddle/phi/kernels/impl/addcmul_kernel_impl.h
paddle/phi/kernels/impl/addcmul_grad_kernel_impl.h
paddle/phi/kernels/cpu/addcmul_kernel.cc
paddle/phi/kernels/cpu/addcmul_grad_kernel.cc
paddle/phi/kernels/gpu/addcmul_kernel.cu
paddle/phi/kernels/gpu/addcmul_grad_kernel.cu
test/legacy_test/test_addcmul.py
```

**Modified Files** (9):
```bash
paddle/phi/ops/yaml/ops.yaml
paddle/phi/ops/yaml/backward.yaml
paddle/phi/infermeta/ternary.h
paddle/phi/infermeta/ternary.cc
paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.h
paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc
python/paddle/tensor/__init__.py
python/paddle/tensor/math.py
test/white_list/op_threshold_white_list.py
```

### 是否引起精度变化
否

### CODE REVIEW

File: python/paddle/tensor/math.py [hunk 1]
```diff
@@ -2331,6 +2331,83 @@ def addmm_(
         return _C_ops.addmm_(input, x, y, beta, alpha)
 
 
+def addcmul(
```
Comment (zhwesky2010) [inline python/paddle/tensor/math.py]: 新增API直接采取C++下沉的方法吧，这个可以不加

**Review by zhwesky2010 — COMMENTED** [review]
看下覆盖率，确保都能测到。

同时提前进行下PaConvert测试，确保与torch计算结果一致。截图下PaConvert的case计算结果。

File: python/paddle/_paddle_docs.py [hunk 2]
```diff
@@ -4187,70 +4187,51 @@ def conj(
 )
 
 add_doc_and_signature(
-    "i1",
+    "addcmul",
```
Comment (zhwesky2010) [inline python/paddle/_paddle_docs.py]: 不要删掉别的，改完后自己先check下所有改动是否符合预期

**Review by zhwesky2010 — COMMENTED** [review]
这个PR看怎么减小下大小

File: python/paddle/tensor/math.py [hunk 3]
```diff
@@ -2333,6 +2334,83 @@ def addmm_(
         return _C_ops.addmm_(input, x, y, beta, alpha)
 
 
+# def addcmul(
```
Comment (zhwesky2010) [inline python/paddle/tensor/math.py]: 这个PR压缩下行数，这些删除掉

File: python/paddle/_paddle_docs.py [hunk 4]
```diff
@@ -4263,3 +4263,49 @@ def i1e(
...
+            >>> x = paddle.to_tensor([1.0, 2.0, 3.0, 4.0])
+            >>> y = paddle.to_tensor([0.1, 0.2, 0.3, 0.4])
+            >>> z = paddle.to_tensor([10.0, 20.0, 30.0, 40.0])
+            >>> result = paddle.addcmul(x, y, z, value=0.1)
+            >>> print(result)
+            Tensor(shape=[4], dtype=float32, place=Place(cpu), stop_gradient=True,
+            [1.10000002, 2.40000010, 3.90000010, 5.60000038])
+""",
+    """
+def addcmul(
```
Comment (zhwesky2010) [inline python/paddle/_paddle_docs.py:4850]: 这个参数名和torch对齐了没，在paconvert上测试了没有

infoflow 2026-01-26 12-23-28

File: paddle/phi/ops/yaml/ops.yaml [hunk 5]
```diff
@@ -7,6 +7,7 @@
 
   # interfaces : paddle::dialect::InferSymbolicShapeInterface
 
+# Trigger rebuild for addcmul
```
Comment (zhwesky2010) [inline paddle/phi/ops/yaml/ops.yaml]: 这个注释不要

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc [hunk 6]
```diff
@@ -62,6 +62,61 @@ bool AccuracyOpInferSymbolicShape(
   return true;
 }
 
+bool AddcmulOpInferSymbolicShape(
```
Comment (zhwesky2010) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc:65]: 这个逻辑在单测里测试到了没，这个需要开CINN才能测到：check_cinn=True

File: test/legacy_test/test_addcmul_op.py [hunk 7]
```diff
@@ -0,0 +1,658 @@
...
+
+    def init_shapes_and_data(self):
+        self.inputs = {
+            'input': np.random.random((10, 20)).astype(self.dtype),
+            'tensor1': np.random.random((10, 20)).astype(self.dtype),
+            'tensor2': np.random.random((10, 20)).astype(self.dtype),
+        }
+        self.attrs = {'value': 0.5}
+
+    def test_check_output(self):
```
Comment (zhwesky2010) [inline test/legacy_test/test_addcmul_op.py:57]: check_cinn测下InferSymbolicShape

File: test/legacy_test/test_addcmul_op.py [hunk 8]
```diff
@@ -0,0 +1,658 @@
...
+        self.attrs = {'value': value}
+        self.outputs = {
+            'out': convert_float_to_uint16(
+                input_np + value * tensor1_np * tensor2_np
+            )
+        }
+
+    def test_check_output(self):
+        place = get_device_place()
+        self.check_output_with_place(place, check_pir=True)
```
Comment (zhwesky2010) [inline test/legacy_test/test_addcmul_op.py:247]: 这里需要额外选择place吗

File: python/paddle/_paddle_docs.py (hunk 4 above)
Comment (zhwesky2010) [inline python/paddle/_paddle_docs.py:4850]: Pytorch的签名

infoflow 2026-01-27 12-40-59

File: test/white_list/op_threshold_white_list.py [hunk 9]
```diff
@@ -56,6 +56,7 @@
     'qr',
     'layer_norm',
     'cumprod',
+    'addcmul',
```
Comment (zhwesky2010) [inline test/white_list/op_threshold_white_list.py:59]: 是否windows的问题？

**Review by zhwesky2010 — COMMENTED** [review]
检查下哪些代码没测到，覆盖率没过

File: python/paddle/_paddle_docs.py (hunk 4 above)
Comment (Manfredss) [reply python/paddle/_paddle_docs.py:4850]: 是需要对齐报错信息吗？

File: test/white_list/op_threshold_white_list.py (hunk 9 above)
Comment (Manfredss) [reply test/white_list/op_threshold_white_list.py:59]: 这个添加是需要的，因为：
1. 测试使用 check_cinn=True，这会引入额外的数值变化
2. addcmul 运算本身 (input + value × tensor1 × tensor2) 就会积累浮点运算误差

File: python/paddle/_paddle_docs.py (hunk 4 above)
Comment (zhwesky2010) [reply python/paddle/_paddle_docs.py:4850]: > 是需要对齐报错信息吗？

不是报错信息，需要对齐签名顺序和名字

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc [hunk 10]
```diff
@@ -62,6 +62,61 @@ bool AccuracyOpInferSymbolicShape(
...
+    pir::Operation *op, pir::InferSymbolicShapeContext *infer_context) {
+  const auto &x_shape_or_data =
```
Comment (zhwesky2010) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc]: 这里的命名改成 input, tensor1, tensor2吧，提升下可读性。这里x/y/z命名无意义。

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc [hunk 11]
```diff
@@ -62,6 +62,61 @@ bool AccuracyOpInferSymbolicShape(
...
+      infer_context->GetShapeOrDataForValue(op->operand_source(2));
+  std::vector<symbol::DimExpr> x_shape = x_shape_or_data.shape();
+  std::vector<symbol::DimExpr> y_shape = y_shape_or_data.shape();
+  std::vector<symbol::DimExpr> z_shape = z_shape_or_data.shape();
+
+  // NOTE(large-tensor): tensor dimensions are small integers
+  int x_ndims = static_cast<int>(x_shape.size());
+  int y_ndims = static_cast<int>(y_shape.size());
+  int z_ndims = static_cast<int>(z_shape.size());
+  std::vector<symbol::DimExpr> out1_shape;
```
Comment (zhwesky2010) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/multiary_infer_sym.cc]: 不用这种a1/a2/a3、a/b/c、x/y/z、out1/out2/out3的命名方式。用实际的功能含义来命名。

File: paddle/phi/kernels/impl/addcmul_grad_kernel_impl.h [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+                               &reduce_vec);
+      ReduceGrad<Context, T, Dims>(
+          dev_ctx, out_grad, reshape_vec, reduce_vec, input_grad);
+    }
+  }
+
+  // d(tensor1) = reduce_sum(dout * value * tensor2)
+  if (tensor1_grad) {
+    DenseTensor temp;
+    temp.Resize(out_grad.dims());
```
Comment (zhwesky2010) [inline paddle/phi/kernels/impl/addcmul_grad_kernel_impl.h]: 这个temp只有broadcase情况下才需要，应该可以节省一个显存+copy成本：
1. 无需broadcase情况下：直接计算tensor1_grad
2. 需broadcast情况下：创建tensor1_grad_b，先计算tensor1_grad_b，然后reduce

然后命名不用temp这些无意义命名，用实际语义命名：tensor1_grad_b表示是tensor1_grad在broadcast后的。

**Review by zhwesky2010 — DISMISSED** [review]
LGTM

**Comment by zhwesky2010** [comment]
@Manfredss 看下Coverage流水线问题

**Comment by YqGe585** [comment]
IXUCA rerun仍然失败的,辛苦rebase到develop后重新触发CI

**Comment by zhwesky2010** [comment]
Coverage和IXUCA失败了

**Comment by zhwesky2010** [comment]
这个Coverage和XPU还有问题吗

**Comment by Manfredss** [comment]
> 这个Coverage和XPU还有问题吗

Coverage 里那几个 file 总是覆盖不到，然后 XPU 感觉和我的改动没有关系？

**Comment by zhwesky2010** [comment]
> > 这个Coverage和XPU还有问题吗
> 
> Coverage 里那几个 file 总是覆盖不到，然后 XPU 感觉和我的改动没有关系？

XPU可以skip过去吗

**Review by zhwesky2010 — DISMISSED** [review]

**Comment by zhwesky2010** [comment]
LGTM

**Comment by zhwesky2010** [comment]
infoflow 2026-02-11 17-11-14

OP单测运行失败了，需要看下。

另外如果整个文件都skip xpu了，可以直接在CMakeLists.txt里改，就不用逐个mark skipif了。

**Comment by Manfredss** [comment]
> infoflow 2026-02-11 17-11-14
> OP单测运行失败了，需要看下。
> 
> 另外如果整个文件都skip xpu了，可以直接在CMakeLists.txt里改，就不用逐个mark skipif了。

这是哪个设备的单测没过？我这看不到
image

**Comment by zhwesky2010** [comment]
infoflow 2026-02-11 18-33-31

Static-Check示例代码没过

**Review by zhwesky2010 — DISMISSED** [review]

**Comment by zhwesky2010** [comment]
示例代码还是没过，是不是要导入到 \__init\__.py里去

infoflow 2026-02-12 15-52-17

**Comment by SigureMo** [comment]
> 示例代码还是没过，是不是要导入到 `__init__.py` 里去

对的，需要的，要在 `paddle/__init__.py` 加一下

**Review by zhwesky2010 — DISMISSED** [review]

**Review by SigureMo — DISMISSED** [review]

File: test/legacy_test/test_addcmul_op.py [hunk 13]
```diff
@@ -0,0 +1,901 @@
...
+        net_cinn = paddle.jit.to_static(
+            addcmul_loss,
+            input_spec=input_specs,
+            backend="CINN",
+            full_graph=True,
+        )
+
+        cinn_loss = net_cinn(x, t1, t2)
+        np.testing.assert_allclose(
+            cinn_loss.numpy(), dy_loss.numpy(), rtol=1e-5, atol=1e-5
```
Comment (zhwesky2010) [inline test/legacy_test/test_addcmul_op.py]: 这几个地方建议精简下case，避免atol/rtol的使用。

File: test/legacy_test/test_addcmul_op.py [hunk 14]
```diff
@@ -0,0 +1,901 @@
...
```
Comment (zhwesky2010) [inline test/legacy_test/test_addcmul_op.py]: 同上

File: test/legacy_test/test_addcmul_op.py [hunk 15]
```diff
@@ -0,0 +1,901 @@
...
+            full_graph=True,
+        )
+        net_cinn.eval()
+        cinn_out = net_cinn(*inputs)
+
+        # Run without CINN (dynamic graph reference)
+        dy_out = fn(*inputs)
+
+        np.testing.assert_allclose(
+            cinn_out.numpy(), dy_out.numpy(), rtol=1e-5, atol=1e-5
```
Comment (zhwesky2010) [inline test/legacy_test/test_addcmul_op.py]: cinn的计算结果与常规的应该是一样的，这里降低阈值的原因是

File: test/legacy_test/test_addcmul_op.py [hunk 16]
```diff
@@ -0,0 +1,901 @@
...
+        out = paddle.addcmul(input_t, tensor1_t, tensor2_t, value=0.5)
+        out.sum().backward()
+
+        self.assertEqual(list(input_t.grad.shape), [0, 4])
+        self.assertEqual(list(tensor1_t.grad.shape), [0, 4])
+        self.assertEqual(list(tensor2_t.grad.shape), [0, 4])
+        paddle.enable_static()
+
+
+@unittest.skipIf(
```
Comment (zhwesky2010) [inline test/legacy_test/test_addcmul_op.py]: 这里如果需要大量skipif，还是从CMakeLists.txt里面来控制吧。需尽量避免skipif的使用。

**Review by zhwesky2010 — COMMENTED** [review]
Coverage无法通过的原因是test_addcmul无法运行通过，本地在GPU上测下addcmul吧。
同时尽可能减少skipif和atol/rtol的修改。

**Comment by zhwesky2010** [comment]
单测还是运行失败了，看起来是静态图测试InferSymbolicShape的case没跑过，你本地运行能跑过吗？

infoflow 2026-03-05 18-52-22

File: test/legacy_test/test_addcmul_op.py [hunk 17]
```diff
@@ -0,0 +1,872 @@
...
+        np.testing.assert_allclose(ref_out, out3.numpy(), rtol=1e-5)
+        paddle.enable_static()
+
+    def test_static_api(self):
+        """Test static graph API"""
+        paddle.enable_static()
+        main = paddle.static.Program()
+        startup = paddle.static.Program()
+        with base.program_guard(main, startup):
+            x = paddle.static.data(name="x", shape=self.shape, dtype=self.dtype)
```
Comment (zhwesky2010) [inline test/legacy_test/test_addcmul_op.py]: 看这个报错信息是 PIR下创建OP时，创建完一个DataOP，在Insert到block时越界错误。看起来和addcmul自身的OP逻辑无关，是不是单测本身不对。本地复现下问题调试看看。

Image

Image

Comment (Manfredss) [reply test/legacy_test/test_addcmul_op.py]: 主要我这边用 python 测单测一直是通过的
image

Comment (zhwesky2010) [reply test/legacy_test/test_addcmul_op.py]: 你有没有Linux GPU的运行环境

**Comment by zhwesky2010** [comment]
@Manfredss 这个PR的问题复现了吗

**Comment by zhwesky2010** [comment]
@Manfredss 这个尽快调试吧

**Comment by Manfredss** [comment]
> @Manfredss 这个PR的问题复现了吗

我 linux 下好像也没问题？
image

**Comment by Manfredss** [comment]
屏幕截图 2026-04-12 110117

开发机上也 ok

File: paddle/phi/infermeta/ternary.cc [hunk 18]
```diff
@@ -139,6 +142,51 @@ void AddmmInferMeta(const MetaTensor& input,
...
+    if (x_val != y_val && x_val != 1 && y_val != 1) {
+      PADDLE_THROW(common::errors::InvalidArgument(
+          "Broadcast dimension mismatch. Operands could not be broadcast "
+          "together with shapes %s and %s",
+          x_dims,
+          y_dims));
+    }
+    out_dims_vector[i] = std::max(x_val, y_val);
+  }
+  return common::make_ddim(out_dims_vector);
```
Comment (risemeup1111) [inline paddle/phi/infermeta/ternary.cc]: P2 **优先级：P2**

这里的私有广播实现和仓库里现有的广播语义不一致，`0` 和 `1` 的组合会被算成 `1`。例如 `input=[0,3]`、`tensor1=[1,3]`、`tensor2=[1,3]` 现在会推成 `[1,3]`，但后续 kernel / 现有广播辅助函数会按 `0` 维处理。这会让 infermeta 和运行时形状校验不一致。

```suggestion
static DDim BroadcastTwoDims(const DDim& x_dims, const DDim& y_dims) {
  int max_rank = std::max(x_dims.size(), y_dims.size());
  std::vector<int64_t> out_dims_vector(max_rank);

  int x_idx = x_dims.size() - 1;
  int y_idx = y_dims.size() - 1;

  for (int i = max_rank - 1; i >= 0; --i) {
    int64_t x_val = x_idx >= 0 ? x_dims[x_idx--] : 1;
    int64_t y_val = y_idx >= 0 ? y_dims[y_idx--] : 1;

    if (x_val == y_val) {
      out_dims_vector[i] = x_val;
    } else if (x_val == 1) {
      out_dims_vector[i] = y_val;
    } else if (y_val == 1) {
      out_dims_vector[i] = x_val;
    } else {
      PADDLE_THROW(common::errors::InvalidArgument(
          "Broadcast dimension mismatch. Operands could not be broadcast "
          "together with shapes %s and %s",
          x_dims,
          y_dims));
    }
  }
  return common::make_ddim(out_dims_vector);
}
```

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

**Review by risemeup1111 — COMMENTED** [review]
已复查，未发现阻塞性问题；有一处中优先级问题，细节见 inline review comments。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddle/phi/infermeta/ternary.cc (hunk 18 above)
Comment (Manfredss) [reply paddle/phi/infermeta/ternary.cc]: Done

**Review by risemeup1111 — APPROVED** [review]
已复查当前提交，先前的 P2 已通过新提交修复并有作者回复；代码层面未发现需要继续阻塞的问题。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddle/phi/infermeta/ternary.cc [hunk 19]
```diff
@@ -139,6 +142,44 @@ void AddmmInferMeta(const MetaTensor& input,
   out->set_dtype(input.dtype());
 }
 
+void AddcmulInferMeta(const MetaTensor& input,
+                      const MetaTensor& tensor1,
+                      const MetaTensor& tensor2,
+                      MetaTensor* out) {
+  PADDLE_ENFORCE_EQ(
+      tensor1.dtype(),
```
Comment (risemeup1111) [inline paddle/phi/infermeta/ternary.cc]: P1 **优先级：P1**

这里强制三个输入的 dtype 完全一致，会拒绝 PyTorch 支持的类型提升用法，例如 `input=float32`、`tensor1/tensor2=float64`。PyTorch `addcmul` 的 TensorIterator 会将输入提升到公共 dtype；而对应 PaConvert 使用 `ChangePrefixMatcher` 后，这类合法调用会在 Paddle 侧直接报错。请让非 inplace API 按公共 dtype 计算（inplace API 再执行安全转换约束），并把当前固定“dtype mismatch 必须报错”的测试改为混合 dtype 对齐测试。

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

File: paddle/phi/kernels/impl/addcmul_kernel_impl.h [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+  PADDLE_ENFORCE_GE(
+      rank,
+      0,
+      common::errors::InvalidArgument(
+          "The number of dimensions for AddcmulOp must be "
+          "greater than or equal to 0, but the value received is %d.",
+          rank));
+  PADDLE_ENFORCE_LE(
+      rank,
+      6,
```
Comment (risemeup1111) [inline paddle/phi/kernels/impl/addcmul_kernel_impl.h]: P1 **优先级：P1**

这个硬限制会让任意 7 维输入（例如三个 shape 都为 `[1, 1, 1, 1, 1, 1, 1]`）在 forward 直接报错，backward 也有同样限制；PyTorch `addcmul` 没有该维数上限，公开文档中也未声明这一差异。底层 `EigenBroadcast` 已实例化到 9 维，因此至少 7 维并非不可支持。请移除/扩展该限制并补充 7 维 forward、backward 对齐用例。

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

File: paddle/phi/kernels/funcs/common_shape.h [hunk 21]
```diff
@@ -129,7 +129,19 @@ static void GetBroadcastDims(const DDim &in_dims,
...
+          in_dims[i],
+          1,
+          common::errors::InvalidArgument(
+              "Tensor dimension mismatch for broadcasting. "
+              "Input dimension %d is %d, output dimension is %d. "
+              "For broadcasting, input dimension must be 1 or equal to output.",
+              i,
+              in_dims[i],
+              out_dims[i]));
+      (*bcast_dims)[i] = out_dims[i];
```
Comment (risemeup1111) [inline paddle/phi/kernels/funcs/common_shape.h]: P1 **优先级：P1**

`out_dims[i]` 是 64 位 shape dim，但这里写入的是 `Eigen::DSizes<int, D>`。当广播维超过 `INT_MAX`（如 `[1]` 广播到 `[2147483648]`）时会发生无范围检查的 64→32 位收窄，forward 得到无效的 broadcast factor。请将该辅助函数及调用方的 `DSizes` 索引类型改为 `int64_t`/`Eigen::DenseIndex`，不要在 shape 路径保留 32 位承接。

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

File: paddle/phi/kernels/impl/addcmul_grad_kernel_impl.h [hunk 22]
```diff
@@ -0,0 +1,348 @@
...
+                                     std::vector<int>* reduce_dims_vec) {
+  std::vector<int64_t> extended = grad_dims;
+  size_t diff = out_dims.size() - grad_dims.size();
+  extended.insert(extended.begin(), diff, 1);
+
+  reshape_dims_vec->clear();
+  reduce_dims_vec->clear();
+  for (size_t i = 0; i < extended.size(); ++i) {
+    reduce_dims_vec->push_back(static_cast<int>(reshape_dims_vec->size()));
+    reshape_dims_vec->push_back(static_cast<int>(out_dims[i] / extended[i]));
```
Comment (risemeup1111) [inline paddle/phi/kernels/impl/addcmul_grad_kernel_impl.h]: P1 **优先级：P1**

这里又把 64 位的广播因子和原始维度显式截成 `int`。例如 `out_grad.shape=[2147483648]`、输入梯度 shape 为 `[1]` 时，reshape factor 无法由 `int` 表示，后续本来使用 `Eigen::DenseIndex` 的 reduction 会收到损坏的 reshape。请让 `reshape_dims_vec` 全程使用 `int64_t`/`Eigen::DenseIndex`；`reduce_dims` 也可保持同一索引类型，避免梯度路径单独留下收窄。

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

File: paddle/phi/kernels/impl/addcmul_kernel_impl.h [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+  if constexpr (std::is_integral<T>::value) {
+    // Align with PyTorch: for integral input tensors a floating-point value
+    // is truncated towards zero, while a value outside the range of the
+    // input data type is rejected.
+    const double numeric_value = value.to<double>();
+    const double lowest = static_cast<double>(std::numeric_limits<T>::lowest());
+    const double highest = static_cast<double>(std::numeric_limits<T>::max());
+    const bool is_within_upper_bound =
+        std::numeric_limits<T>::digits >= std::numeric_limits<double>::digits
+            ? numeric_value < highest
```
Comment (risemeup1111) [inline paddle/phi/kernels/impl/addcmul_kernel_impl.h]: P2 **优先级：P2**

合法的 `int64` 上界会被这里误判为溢出：`INT64_MAX` 转成 `double` 后与 `static_cast<double>(INT64_MAX)` 都变为 `2^63`，因此严格 `< highest` 为 false。比如 `input=0`、`tensor1=tensor2=1`、`value=2**63-1` 的结果本应正好为 `INT64_MAX`，当前却报错，也与文档“within the range”不符。请按 `Scalar` 的原始 dtype 做精确范围判断，并补充两个整数端点用例。

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| - | - | - | - |
| 1 | [`paddle/phi/infermeta/ternary.cc`](https://github.com/PaddlePaddle/Paddle/pull/77333#discussion_r3670692821) | P2 | ✅ |
| 2 | [`paddle/phi/infermeta/ternary.cc`](https://github.com/PaddlePaddle/Paddle/pull/77333#discussion_r3684531148) | P1 | 🚧 |
| 3 | [`paddle/phi/kernels/impl/addcmul_kernel_impl.h`](https://github.com/PaddlePaddle/Paddle/pull/77333#discussion_r3684534993) | P1 | ✅ |
| 4 | [`paddle/phi/kernels/funcs/common_shape.h`](https://github.com/PaddlePaddle/Paddle/pull/77333#discussion_r3684537495) | P1 | 🚧 |
| 5 | [`paddle/phi/kernels/impl/addcmul_grad_kernel_impl.h`](https://github.com/PaddlePaddle/Paddle/pull/77333#discussion_r3684540117) | P1 | ✅ |
| 6 | [`paddle/phi/kernels/impl/addcmul_kernel_impl.h`](https://github.com/PaddlePaddle/Paddle/pull/77333#discussion_r3684542536) | P2 | 🚧 |
| 7 | [PR 描述](https://github.com/PaddlePaddle/Paddle/pull/77333) | P3 | 🚧 |

P3 **优先级：P3**

PR 描述的测试文件名仍写为 `test/legacy_test/test_addcmul.py`，但当前 diff 实际新增的是 `test/legacy_test/test_addcmul_op.py`。请更新文件清单及测试命令，避免验收依据指向不存在的路径。

**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddle/phi/infermeta/ternary.cc (hunk 19 above)
Comment (risemeup1111) [reply paddle/phi/infermeta/ternary.cc]: 当前回退仅移除了 InferMeta 的显式 dtype 相等检查，未实现类型提升：`eager_gen.py` 的 `type_promote_white_list` 仍无 `addcmul`，`ops.yaml` 仍以 `input` 选择 kernel；随后 kernel 用同一 `T` 读取 `tensor1`/`tensor2`，混合 dtype 会在 `DenseTensor::data<T>()` 处因类型不符报错。该 P1 仍未解决，请按原建议实现公共 dtype 提升并补充对齐测试。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: paddle/phi/kernels/impl/addcmul_kernel_impl.h (hunk 20 above)
Comment (risemeup1111) [reply paddle/phi/kernels/impl/addcmul_kernel_impl.h]: 已在当前提交验证：forward/backward 已补齐 rank 7-9 分派，并增加 rank 7/9 的前向与梯度覆盖；该项已修复。

File: paddle/phi/kernels/funcs/common_shape.h (hunk 21 above)
Comment (risemeup1111) [reply paddle/phi/kernels/funcs/common_shape.h]: 当前提交仍未修复前向路径：`AddcmulFunction` 的三个 broadcast dims 仍为 `Eigen::DSizes<int, D>`，`GetBroadcastDims` 也仍接收该类型，因此超过 `INT_MAX` 的广播因子仍会收窄。该 P1 保持未解决。

File: paddle/phi/kernels/impl/addcmul_grad_kernel_impl.h (hunk 22 above)
Comment (risemeup1111) [reply paddle/phi/kernels/impl/addcmul_grad_kernel_impl.h]: 已在当前提交验证：reshape/reduce vector 和 `Eigen::DSizes` 均改为 `int64_t`，已移除反向 shape 路径的 32 位收窄；该项已修复。

File: paddle/phi/kernels/impl/addcmul_kernel_impl.h (hunk 23 above)
Comment (risemeup1111) [reply paddle/phi/kernels/impl/addcmul_kernel_impl.h]: 当前提交只修复了动态图端点。`ops.yaml` 仍将 `value` 声明为 `Scalar(double)`，静态/PIR 构图会先把 `2**63-1` 转为 `double`（即 `2**63`），随后仍被范围检查拒绝；新增端点测试也只覆盖 `disable_static` 路径。请保留原始 Scalar 类型并补充静态/PIR 端点用例。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
